### PR TITLE
[Bash] More command formatting

### PR DESCRIPTION
**Pull request by Xophmeister**
This PR introduces queries that improve support for command, command lists and command pipelines, in aid of #138. It is outputting valid code, but suffers from some (non-blocking, imo) shortfalls:

* The line-spacing is not perfect;
* Multi-line "commands" are collapsed into a single line (see #172)

I have also cleaned-up and checked-in the script -- with documentation -- I created to support interactive query development in the terminal (as a quick-and-dirty alternative to #171); it seems generally useful.

### CODE REVIEW

File: languages/bash.scm [hunk 1]
```diff
@@ -5,15 +5,17 @@
...
   (comment)
   (if_statement)
+  (command)
+  (list)
+  (pipeline)
+  ; TODO: etc.
 ] @allow_blank_line_before
 
 ; Prepend spaces
 ; (Not "surround", to avoid extraneous spaces between delimiters)
```
Comment (nbacquey) [inline languages/bash.scm:18]: I think we can surround, and let the post-processing take care of extraneous spaces

File: languages/bash.scm [hunk 2]
```diff
@@ -22,23 +24,43 @@
...
+; collapsed on to a single line for now... See Issue #172
+
+; One command per line in the following contexts:
+; * Top-level
+; * In any branch of a conditional
+; * <TODO: etc.>
+;
+; NOTE Because "command" is such a pervasive and general concept, each
+; context needs to be individually enumerated to account for exceptions.
+(program [(command) (list) (pipeline)] @prepend_hardline)
```
Comment (nbacquey) [inline languages/bash.scm:51]: Even if it's more space-consuming, I think I prefer the heavily-indented way to write queries, even simple ones.
At least for consistency within the query files, e.g.
```
(program
  [
    (command)
    (list)
    (pipeline)
  ] @prepend_hardline
)
```
Or at least
```
(program
  [(command) (list) (pipeline)] @prepend_hardline
)
```
if the `[(command) (list) (pipeline)]` syntagm is expected to be everywhere

**Review by nbacquey — APPROVED** [review]

File: languages/bash.scm (hunk 1 above)
Comment (Xophmeister) [reply languages/bash.scm:18]: The reason this is in here is to protect the `<CLAUSE>; <CLAUSE>` case; i.e., no space before the delimiter. I don't believe the post-processing deals with this.

That said, I've fixed this in #174 by making the "surround with spaces" list much more restrictive.

File: languages/bash.scm (hunk 2 above)
Comment (Xophmeister) [reply languages/bash.scm:51]: Good idea :+1: I'll fix this in #174 to avoid making merge conflict problems for myself

File: languages/bash.scm [hunk 3]
```diff
@@ -22,23 +24,43 @@
...
 ;
-; FIXME Can this be generalised, or does *every* context need to be
-; individually enumerated?...
-(program (command) @prepend_spaced_softline @append_hardline)
-(if_statement _ "then" (command) @prepend_spaced_softline)
-(elif_clause _ "then" (command) @prepend_spaced_softline)
-(else_clause (command) @prepend_spaced_softline)
+; That is: [(command) (list) (pipeline)], per the grammar
+
+; FIXME I don't think it's possible to insert the necessary line
```
Comment (torhovland) [inline languages/bash.scm:39]: Should `command` be treated as a leaf (so we leave it entirely alone)?

File: languages/bash.scm [hunk 4]
```diff
@@ -22,23 +24,43 @@
...
+;
+; NOTE Because "command" is such a pervasive and general concept, each
+; context needs to be individually enumerated to account for exceptions.
+(program [(command) (list) (pipeline)] @prepend_hardline)
+(if_statement _ "then" [(command) (list) (pipeline)] @prepend_hardline)
+(elif_clause _ "then" [(command) (list) (pipeline)] @prepend_hardline)
+(else_clause [(command) (list) (pipeline)] @prepend_hardline)
+
+; Surround command list and pipeline delimiters with spaces
+(list ["&&" "||"] @append_space @prepend_space)
```
Comment (torhovland) [inline languages/bash.scm:57]: Could we just put these in the list of symbols to surround with spaces? Do they really need to come within `list` and `pipeline`?

Comment (Xophmeister) [reply languages/bash.scm:57]: I guess they don't. I've done it like this to make the context explicit, as I build the queries up incrementally; but you're right that it's redundant.

I'll add a note next to these rules (in #174) to ultimate generalise them into the list of symbols to surround with spaces.

Comment (Xophmeister) [reply languages/bash.scm:57]: Note added in 6b4c69e8641193

File: languages/bash.scm (hunk 3 above)
Comment (Xophmeister) [reply languages/bash.scm:39]: I'm not fully clear about the consequences of `@leaf`. I understand that Topiary just skips over these nodes, but to what end/what effect would that have on `(command)`?

Comment (torhovland) [reply languages/bash.scm:39]: It would keep leafs formatted as is. I would think that might be better than forcing all commands to be single-line.

Comment (Xophmeister) [reply languages/bash.scm:39]: Ah, I see...

So the following:

```bash
foo \
--bar
```

Will be formatted as `foo --bar`, without making it a leaf. Whereas, as a leaf, like you say, no formatting changes are applied. That's certainly better, assuming that's what the author intended :+1:

Changed in cfbd8cb2cd64568

Comment (Xophmeister) [reply languages/bash.scm:39]: ...Oh, but then that obviates this rule:

```scheme
; Space between command line arguments
(command
  argument: _* @append_space @prepend_space
)
```

Which formats, e.g., `foo     --bar  --quux` into `foo --bar --quux`... I quite like this rule.

Comment (torhovland) [reply languages/bash.scm:39]: Yeah, I see.

**Pull request merged**